### ISSUE TITLE
[Bug] - `amazon-chrony-config` breaks `cloud-init`

### ISSUE THREAD

**Issue by adrianmace** [amazonlinux/amazon-linux-2023#345]
**Describe the bug**
If `chrony-config.service` runs before `cloud-init.service` runs **for the first time**, `chrony-config.service` will place a file at the location `/var/lib/cloud/instance/sem/` and sets up the directory structure, since it does not exist yet.

This causes `cloud-init.service` to fail to start, as it expects `/var/lib/cloud/instance` to be a symlink, not a directory.

```shell
[...]
[    6.129038] RPC: Registered udp transport module.
[    6.129546] RPC: Registered tcp transport module.
[    6.130055] RPC: Registered tcp NFSv4.1 backchannel transport module.
[    6.713143] ena 0000:00:05.0 ens5: Local page cache is disabled for less than 16 channels
[    7.735378] cloud-init[1572]: Cloud-init v. 22.2.2 running 'init' at Thu, 11 May 2023 01:17:04 +0000. Up 7.67 seconds.
[    7.842043] cloud-init[1572]: ci-info: ++++++++++++++++++++++++++++++++++++++Net device info++++++++++++++++++++++++++++++++++++++
[    7.844002] cloud-init[1572]: ci-info: +--------+------+----------------------------+---------------+--------+-------------------+
[    7.845712] cloud-init[1572]: ci-info: | Device |  Up  |          Address           |      Mask     | Scope  |     Hw-Address    |
[    7.847254] cloud-init[1572]: ci-info: +--------+------+----------------------------+---------------+--------+-------------------+
[    7.860386] cloud-init[1572]: ci-info: |  ens5  | True |        10.49.28.17         | 255.255.255.0 | global | 02:ec:89:ce:26:d2 |
[    7.862158] cloud-init[1572]: ci-info: |  ens5  | True | fe80::ec:89ff:fece:26d2/64 |       .       |  link  | 02:ec:89:ce:26:d2 |
[    7.863836] cloud-init[1572]: ci-info: |   lo   | True |         127.0.0.1          |   255.0.0.0   |  host  |         .         |
[    7.865339] cloud-init[1572]: ci-info: |   lo   | True |          ::1/128           |       .       |  host  |         .         |
[    7.866834] cloud-init[1572]: ci-info: +--------+------+----------------------------+---------------+--------+-------------------+
[    7.868366] cloud-init[1572]: ci-info: +++++++++++++++++++++++++++++Route IPv4 info++++++++++++++++++++++++++++++
[    7.869730] cloud-init[1572]: ci-info: +-------+-------------+------------+-----------------+-----------+-------+
[    7.871096] cloud-init[1572]: ci-info: | Route | Destination |  Gateway   |     Genmask     | Interface | Flags |
[    7.872503] cloud-init[1572]: ci-info: +-------+-------------+------------+-----------------+-----------+-------+
[    7.873879] cloud-init[1572]: ci-info: |   0   |   0.0.0.0   | 10.49.28.1 |     0.0.0.0     |    ens5   |   UG  |
[    7.875252] cloud-init[1572]: ci-info: |   1   |  10.49.28.0 |  0.0.0.0   |  255.255.255.0  |    ens5   |   U   |
[    7.876627] cloud-init[1572]: ci-info: |   2   |  10.49.28.1 |  0.0.0.0   | 255.255.255.255 |    ens5   |   UH  |
[    7.878036] cloud-init[1572]: ci-info: |   3   |  10.49.28.2 |  0.0.0.0   | 255.255.255.255 |    ens5   |   UH  |
[    7.879515] cloud-init[1572]: ci-info: +-------+-------------+------------+-----------------+-----------+-------+
[    7.881038] cloud-init[1572]: ci-info: +++++++++++++++++++Route IPv6 info+++++++++++++++++++
[    7.882213] cloud-init[1572]: ci-info: +-------+-------------+---------+-----------+-------+
[    7.883356] cloud-init[1572]: ci-info: | Route | Destination | Gateway | Interface | Flags |
[    7.884505] cloud-init[1572]: ci-info: +-------+-------------+---------+-----------+-------+
[    7.885658] cloud-init[1572]: ci-info: |   1   |  fe80::/64  |    ::   |    ens5   |   U   |
[    7.886816] cloud-init[1572]: ci-info: |   3   |    local    |    ::   |    ens5   |   U   |
[    7.888011] cloud-init[1572]: ci-info: |   4   |  multicast  |    ::   |    ens5   |   U   |
[    7.889172] cloud-init[1572]: ci-info: +-------+-------------+---------+-----------+-------+
[    7.890323] cloud-init[1572]: 2023-05-11 01:17:04,287 - util.py[WARNING]: failed stage init
[    7.891433] cloud-init[1572]: failed run of stage init
[    7.892158] cloud-init[1572]: ------------------------------------------------------------
[    7.893269] cloud-init[1572]: Traceback (most recent call last):
[    7.894190] cloud-init[1572]:   File "/usr/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 740, in status_wrapper
[    7.895824] cloud-init[1572]:     ret = functor(name, args)
[    7.896685] cloud-init[1572]:   File "/usr/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 362, in main_init
[    7.898153] cloud-init[1572]:     init.fetch(existing=existing)
[    7.910571] cloud-init[1572]:   File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 430, in fetch
[    7.912074] cloud-init[1572]:     return self._get_data_source(existing=existing)
[    7.913780] cloud-init[1572]:   File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 317, in _get_data_source
[    7.915224] cloud-init[1572]:     util.del_file(self.paths.instance_link)
[    7.916149] cloud-init[1572]:   File "/usr/lib/python3.9/site-packages/cloudinit/util.py", line 1967, in del_file
[�[0;1;31mFAILED�[0m] Failed to start �[0;1;39mcloud-initâ�¦it job (metadata service crawler).
[    7.923340] cloud-init[1572]:     raise e
[    7.927461] cloud-init[1572]:   File "/usr/lib/python3.9/site-packages/cloudinit/util.py", line 1964, in del_file
[    7.928955] cloud-init[1572]:     os.unlink(path)
[    7.929719] cloud-init[1572]: IsADirectoryError: [Errno 21] Is a directory: '/var/lib/cloud/instance'
[    7.931007] cloud-init[1572]: ------------------------------------------------------------
[�[0;1;31mFAILED�[0m] Failed to start �[0;1;39msshd.service�[0m - OpenSSH server daemon.
[    8.374435] cloud-init[1579]: failed run of stage modules-config
[    8.375560] cloud-init[1579]: ------------------------------------------------------------
[    8.377660] cloud-init[1579]: Traceback (most recent call last):
[    8.378729] cloud-init[1579]:   File "/usr/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 740, in status_wrapper
[    8.380380] cloud-init[1579]:     ret = functor(name, args)
[    8.381236] cloud-init[1579]:   File "/usr/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 556, in main_modules
[    8.395299] cloud-init[1579]:     init.fetch(existing="trust")
[    8.396268] cloud-init[1579]:   File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 430, in fetch
[    8.397755] cloud-init[1579]:     return self._get_data_source(existing=existing)
[    8.398805] cloud-init[1579]:   File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 317, in _get_data_source
[    8.400245] cloud-init[1579]:     util.del_file(self.paths.instance_link)
[    8.401195] cloud-init[1579]:   File "/usr/lib/python3.9/site-packages/cloudinit/util.py", line 1967, in del_file
[    8.402583] cloud-init[1579]:     raise e
[    8.403218] cloud-init[1579]:   File "/usr/lib/python3.9/site-packages/cloudinit/util.py", line 1964, in del_file
[    8.404607] cloud-init[1579]:     os.unlink(path)
[    8.405302] cloud-init[1579]: IsADirectoryError: [Errno 21] Is a directory: '/var/lib/cloud/instance'
[    8.406553] cloud-init[1579]: ------------------------------------------------------------
[�[0;1;31mFAILED�[0m] Failed to start �[0;1;39mcloud-confâ�¦ettings specified in cloud-config.
[    8.849973] cloud-init[1596]: failed run of stage modules-final
[    8.851049] cloud-init[1596]: ------------------------------------------------------------
[    8.853703] cloud-init[1596]: Traceback (most recent call last):
[    8.854621] cloud-init[1596]:   File "/usr/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 740, in status_wrapper
[    8.856129] cloud-init[1596]:     ret = functor(name, args)
[    8.875688] cloud-init[1596]:   File "/usr/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 556, in main_modules
[    8.877360] cloud-init[1596]:     init.fetch(existing="trust")
[    8.880176] cloud-init[1596]:   File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 430, in fetch
[    8.897724] cloud-init[1596]:     return self._get_data_source(existing=existing)
[    8.898868] cloud-init[1596]:   File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 317, in _get_data_source
[    8.908875] cloud-init[1596]:     util.del_file(self.paths.instance_link)
[    8.909879] cloud-init[1596]:   File "/usr/lib/python3.9/site-packages/cloudinit/util.py", line 1967, in del_file
[    8.911338] cloud-init[1596]:     raise e
[�[0;1;31mFAILED�[0m] Failed to start �[0;1;39mcloud-finaâ�¦ Execute cloud user/final scripts.
[    8.939696] cloud-init[1596]:   File "/usr/lib/python3.9/site-packages/cloudinit/util.py", line 1964, in del_file
[    8.942406] cloud-init[1596]:     os.unlink(path)
[    8.943485] cloud-init[1596]: IsADirectoryError: [Errno 21] Is a directory: '/var/lib/cloud/instance'
[    8.945155] cloud-init[1596]: ------------------------------------------------------------
2023/05/11 01:17:06Z: Amazon SSM Agent v3.1.1927.0 is running
2023/05/11 01:17:06Z: OsProductName: Amazon Linux
2023/05/11 01:17:06Z: OsVersion: 2023
[...]
```

**To Reproduce**
1. Launch an AL2023 EC2 instance
2. `dnf install -y amazon-chrony-config`
3. `systemctl stop chrony-config.service`
4. Run cleanup script as used by AWS (specifically [these lines](https://github.com/aws/amazon-ecs-ami/blob/6ce3d8b331f5bf0229b47bc0cb5f0d42f5eef51c/scripts/cleanup.sh#L20-L36))
5. Reboot instance
6. Observe that instance is not connectable, since SSH is not available and instance log shows the above messages for each stage of cloud-init.

**Expected behavior**
We expect cloud-init to complete successfully when amazon-chrony-config package is installed.

Ideally, `chrony-config.service` should start only after the `cloud-init.service` has _completed_, or at least check to see if the symlink has been created before trying to store files within that location.

We have currently implemented a workaround by patching the `chrony-config.service` unit file to include an `ExecStartPre=` field that points to a shell script. This shell script tests for the `/var/lib/cloud/instance/boot-finished` file before letting the service start, but this is fragile and we don't want to go into production with it.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by nmeyerhans** [comment]
Yes, the fix for this is queued for inclusion in the next AL2023 release.

**Comment by nmeyerhans** [comment]
This was fixed with the 20230517 release, recently shipped.
